### PR TITLE
Multiple top bottom masks

### PR DESCRIPTION
**Pull request by dham**
Enable multiple different masks in order to support geometric boundary conditions on extruded domains.

### CODE REVIEW

**Comment by miklos1** [comment]
I think this is OK, except that it seems to break the PyOP2 test suite.

**Comment by wence-** [comment]
I think https://github.com/OP2/PyOP2/commit/d4eb085d977035eb8d134955d58e11a5cdeb7633 needs merging into master

File: pyop2/host.py [hunk 1]
```diff
@@ -538,16 +538,22 @@ def c_map_bcs(self, sign):
...
-                    continue
-                need_bottom = True
-                for idx in range(m.arity):
-                    if m.bottom_mask[idx] < 0:
-                        val.append("xtr_%(name)s[%(ind)s] %(sign)s= %(val)s;" %
-                                   {'name': self.c_map_name(i, j),
-                                    'val': max_int,
-                                    'ind': idx,
-                                    'sign': sign})
+                bottom_masks = None
```
Comment (kynan) [inline pyop2/host.py]: Wouldn't it make more sense for this to start as an empty list?

Comment (miklos1) [inline pyop2/host.py]: I agree.

File: pyop2/host.py [hunk 2]
```diff
@@ -559,18 +565,25 @@ def c_map_bcs(self, sign):
...
-                    continue
-                need_top = True
-                for idx in range(m.arity):
-                    if m.top_mask[idx] < 0:
-                        val.append("xtr_%(name)s[%(ind)s] %(sign)s= %(val)s;" %
-                                   {'name': self.c_map_name(i, j),
-                                    'val': max_int,
-                                    'ind': idx,
-                                    'sign': sign})
+                top_masks = None
```
Comment (miklos1) [inline pyop2/host.py]: Same as above.

**Comment by miklos1** [comment]
I think this is mostly okay, although GitHub says there is a merge conflict.

File: pyop2/host.py (hunk 1 above)
Comment (dham) [inline pyop2/host.py]: I disagree. Bottom_masks is never a list. When it's not None it's a numpy array.

Comment (miklos1) [inline pyop2/host.py]: Oh, I didn't realise that. In that case I think `np.zeros` could be a sensible initial value.

Comment (dham) [inline pyop2/host.py]: You could do that, but then the "if bottom_masks ..." below becomes a requirement to check that every entry in the mask is zero. I'm not convinced that's much of an improvement overall.

Comment (miklos1) [inline pyop2/host.py]: I wonder if the current version might potentially be dangerous. Line 541 seems to copy a reference to a numpy array, while 543 seems to modify that.

Comment (dham) [inline pyop2/host.py]: Ouch! Nice catch.
